blog/posts/2024-11-07-how-to-get-a-rust-pkg-to-cran/index.qmd

```diff
@@ -0,0 +1,207 @@
+The easiest way is to use the packages [`devtools`](https://devtools.r-lib.org/) and [`usethis`](https://usethis.r-lib.org/):
+
+```r
+# create the basic folder structure of a package
+devtools::create("myRpkg")
+# make sure the following are executed from the new package
+setwd("myRpkg")
+# set license to MIT
+usethis::use_mit_license()
+# use RMarkdown for Readme
+usethis::use_readme_rmd()
+# use NEWS.md
+usethis::use_news_md()
+# use cran-comments.md - will be important later
+usethis::use_cran_comments()
+```
+
+And with this we should have the basic R package.
+
+A little bit of foreshadowing, but we will have to edit our `DESCRIPTION` file and add the right level of detail for our package, such as author, description, URLs etc.
+
+
+### 2. Add Rust as a Dependency
+
+Similar to the `usethis` package, there is the `rextendr` package that makes this step pretty straight forward.
+
+```r
+rextendr::use_extendr()
+```
+
+This will create the required files in `src/` and `src/rust`.
+
+As the command tells us, whenever we update our Rust code, we should run the following to document the code and build the Rust-parts.
+
+```r
+rextendr::document()
+# if we have changed our R-code and its documentation
+# we need the following as well
+devtools::document()
+```
+
+And we should be ready to go and call our default Rust function `hello_world()` (defined in `src/rust/src/lib.rs`).
+
+The actual R and Rust functions are typically the easiest parts of developing a package.
+If you need a good starter, have a look here, eg [`R/get_tokens.R`](https://github.com/DavZim/rtiktoken/blob/master/R/get_tokens.R) as well as [`src/rust/src/lib.rs`](https://github.com/DavZim/rtiktoken/blob/master/src/rust/src/lib.rs) (as we can see, I didn't lie when I said it's a *light* wrapper...).
+
+If we need to add a Rust dependency, we can use `rextendr::use_crate()` or use `cargo add xyz` directly from the `src/rust` directory.
+
+Now on to the "hard" parts.
+
+
+### 3. Get the Package to CRAN
+
+First, we need to make sure that the usual hurdles are met, see also the [R Packages (2e) Book](https://r-pkgs.org/).
+
+- document our functions using [`roxygen2`](https://roxygen2.r-lib.org/) and create the documentation using `devtools::document()`
+- fill the details of our `DESCRIPTION` file, write the `README.Rmd` and knit to `README.md`
+- use [`testthat`](https://testthat.r-lib.org/) and write tests (not strictly needed, but will most likely safe us in the future!)
+- ... other steps that are typically done in R package development
+- make sure `devtools::check()` works without a NOTE
+
+There are however a couple of CRAN-specific rules and best practices for packages using Rust (see also [Using Rust in CRAN Packages](https://cran.r-project.org/web/packages/using_rust.html)).
+Most of these requirements are already met, but there are a couple of must-haves and nice-to-haves.
```

Comment:
Maybe we can list the requirements here briefly. 

Off the top of the head: 

- Rust dependencies need to be vendored (included) in the R package
- The rust and cargo versions must be reported before building the package
- Ensure that the minimum supported version of rust (MSRV) is available
- Use a maximum of 2 threads to build the package